### ISSUE TITLE
ID Token disappears from localstorage upon expiration okta-react

### ISSUE THREAD

**Issue by trentinthomas** [okta/okta-oidc-js#842]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [x] okta-react
- [ ] okta-react-native
- [ ] okta-vue

## I'm submitting a:

- [x] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior
The ID Token disappears after an hour of idle time within our app. We need the ID token to provide logout services in our application.

## Expected behavior
The ID Token should auto-refresh since the okta session is still alive.

## Minimal reproduction of the problem with instructions
Retrieve the id token/access token from OKTA at the implicit callback, then wait the hour long for the ID token to expire.

## Extra information about the use case/user story you are trying to implement
I find it weird that the ID token just disappears from localstorage. Maybe this is intended but I figured when it goes to refresh the tokens it would pick up the ID token as well.

## Environment

- Package Version: 3.0.1
- Browser: Chrome
- OS: Ubuntu 18.04
- Node version (`node -v`): v12.18.1
- Other:

**Comment by swiftone** [comment]
@trentinthomas - Thanks for the report.

Expired/invalid tokens are expected to be removed from localstorage.  Tokens that remain valid are NOT expected to be removed.  There may be a complication if, say, the renewal of one expired token fails while the other token could still be viable.

Are you saying that the access_token refresh is successful, but the id_token is removed from localstorage regardless?

**Comment by trentinthomas** [comment]
@swiftone yes, the access_token seems to refresh successfully but the id_token gets removed, even if they are set to expire at the same time.

A team member said he upgraded locally to the 3.0.3 version of the codebase just to see if things got fixed, and it seems that it fixes our problem here. 🤞 Is there a scheduled time that a 3.0.3 package will be pushed to the npm repository?

**Comment by trentinthomas** [comment]
If it helps, we seem to only see this behavior when the accessToken TTL is 24 hours. Not sure why/if this does actually make a difference.

**Comment by cfbao** [comment]
I have a related issue, where occasionally (for reasons unclear yet) access token may be absent but id token still present, then access token won't be auto-renewed.

**Comment by swiftone** [comment]
@trentinthomas - okta-react 3.0.3 is out on npm, let me know if that helps.

**Comment by d-gottlieb** [comment]
@swiftone - Yesterday we experienced an outage due to this problem and were running 3.0.2. We upgraded to 3.0.3 and the issue seemed to go away. However, 24 hours later (which would make sense if something got missed here), the issue began occurring again for everyone involved almost at exactly the same time. Looking at `okta-token-storage`, only the `access_token` exists and no `id_token`. 

Deleting the `okta-token-storage` and `okta-cache-storage`, then forcing them to be re-issued, seems to fix the problem. Is there something else that possibly got missed in the 3.0.3 PR?

**Comment by swiftone** [comment]
@d-gottlieb - Not so much "missed", as in a dependency.  We have changes in the works in okta-auth-js, and then these changes need to be consumed by okta-react, so expect updates to both Soon :Tm:

**Comment by shuowu** [comment]
@d-gottlieb I think you have session expiration time set as 24hrs for your application, then when session expired, renew process will trigger `error` event to call `onSessionExpired` callback (default to login, or provided customized callback from the config), which may end up with the state you described in the comment.
We are working on a fix for the session expiration issue in React. https://github.com/okta/okta-oidc-js/pull/848/

**Comment by aarongranick-okta** [comment]
@trentinthomas @cfbao If the session is not expired, a possible source of token renew failure is 3rd party cookies being blocked by the browser. 3rd party cookies are required to renew tokens, since it uses a hidden iframe loading an Okta domain. Default settings of Safari 13+ will block all 3rd party cookies by default, and Chrome can be configured to block 3rd party cookies as well. If possible, please try to determine if this could be the underlying issue. We can help provide workarounds if it is not possible to enable 3rd party cookies on your user's browser.

**Comment by aarongranick-okta** [comment]
@trentinthomas @cfbao @d-gottlieb If your app expects **both** the accessToken and the idToken, you might want to provide a custom `isAuthenticated` function. This function will be called by `updateAuthState` and determines if the user is authenticated based on token state. By default, this will return true if **either** accessToken or idToken is valid: https://github.com/okta/okta-oidc-js/blob/master/packages/okta-react/src/AuthService.js#L143

You may want to customize so that it will return true if **both** accessToken and idToken are valid.  This can be done by passing a function in config. This is probably easier if you initialize the AuthService outside the <Security> component as shown in the example here: https://github.com/okta/okta-oidc-js/tree/master/packages/okta-react#example-with-authservice-object

```
var authService;
var config = {
  issuer: ...,
  clientId: ....,
  isAuthenticated: async () => {
   const tokenManager = authService._oktaAuth.tokenManager;
   const accessToken = await tokenManager.get('accessToken');
   const idToken = await tokenManager.get('idToken');
   return (accessToken && idToken);
 },
 onSessionExpired: () => {} // deprecated function
};
authService = new AuthService(config);
```
I also recommend setting `onSessionExpired` to an empty function. We have seen issues with this function and are planning to remove it in the next major release. It is best if a new auth flow is initiated by your own code seeing `isAuthenticated` is false in the authState.

**Comment by cfbao** [comment]
@aarongranick-okta Thanks for the tip. Is the documentation in this [readme](https://github.com/okta/okta-oidc-js/tree/master/packages/okta-react#user-content-configuration-options:~:text=By%20default%2C%20authService%20will%20consider%20a,getIdToken()%20and%20getAccessToken()%20return%20a%20value.) inaccurate then?

> By default, authService will consider a user authenticated if both getIdToken() and getAccessToken() return a value.

I have a workaround right now to check access token is truthy before use, otherwise login, and that's been working for us.

On using a custom `isAuthenticated` function, I considered this before, and maybe it's just my poor JS skill, but I don't see an easy way to check tokens from a custom function, because when it's called, no arguments are passed in, and the caller isn't auth service itself either:
https://github.com/okta/okta-oidc-js/blob/c5412939b49a30aa6335b1d9e6d2605693e0b9d9/packages/okta-react/src/AuthService.js#L143

**Comment by swiftone** [comment]
> Is the documentation in this readme inaccurate then?

It is...for now. Sorry for the confusion.

> I don't see an easy way to check tokens from a custom function, because 

It does take some careful work, but the example @aarongranick-okta listed above with a custom `isAuthenticated` should work - notice how it references the `authService`  variable that is defined BEFORE the `isAuthenticated` function is defined, but that variable is assigned AFTER the `isAuthenticated` is defined.   This works because the function references the variable, but doesn't use the VALUE until it is run, and by then `authService` is properly assigned.

**Comment by cfbao** [comment]
@swiftone I was gonna say my difficulty was the function to the `Security` component, but then I noticed I could just pass an entire `authService` to `Security`... So that should work. Thanks!

**Comment by ChuckJonas** [comment]
Can we please get complete steps to resolve this issue from the Okta team?  This issue is a year old and still impacting the library...
